libmobilecoin/src/transaction.rs

```diff
@@ -441,44 +442,19 @@ pub extern "C" fn mc_transaction_builder_add_output(
 /// * `LibMcError::InvalidInput`
 #[no_mangle]
 pub extern "C" fn mc_transaction_builder_add_output_with_fog_hint_address(
-    transaction_builder: FfiMutPtr<McTransactionBuilder>,
-    amount: u64,
-    recipient_address: FfiRefPtr<McPublicAddress>,
-    fog_hint_address: FfiRefPtr<McPublicAddress>,
-    rng_callback: FfiOptMutPtr<McRngCallback>,
-    out_tx_out_confirmation_number: FfiMutPtr<McMutableBuffer>,
-    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+    _transaction_builder: FfiMutPtr<McTransactionBuilder>,
+    _amount: u64,
+    _recipient_address: FfiRefPtr<McPublicAddress>,
+    _fog_hint_address: FfiRefPtr<McPublicAddress>,
+    _rng_callback: FfiOptMutPtr<McRngCallback>,
+    _out_tx_out_confirmation_number: FfiMutPtr<McMutableBuffer>,
+    _out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
 ) -> FfiOptOwnedPtr<McData> {
     // FIXME(chris): The SDK should probably stop binding to this function, I don't
     // believe that there is legitimate use for this.
     // It should bind "add_change_output" instead.
     // Please speak to me if you disagree.
-    ffi_boundary_with_error(out_error, || {
-        let transaction_builder = transaction_builder
-            .into_mut()
-            .as_mut()
-            .expect("McTransactionBuilder instance has already been used to build a Tx");
-        let recipient_address =
-            PublicAddress::try_from_ffi(&recipient_address).expect("recipient_address is invalid");
-        let fog_hint_address =
-            PublicAddress::try_from_ffi(&fog_hint_address).expect("fog_hint_address is invalid");
-        let mut rng = SdkRng::from_ffi(rng_callback);
-        let out_tx_out_confirmation_number = out_tx_out_confirmation_number
-            .into_mut()
-            .as_slice_mut_of_len(TxOutConfirmationNumber::size())
-            .expect("out_tx_out_confirmation_number length is insufficient");
-
-        let (tx_out, confirmation) = transaction_builder.add_output_with_fog_hint_address(
-            amount,
-            &recipient_address,
-            &fog_hint_address,
-            |_| Ok(MemoPayload::default()),
-            &mut rng,
-        )?;
-
-        out_tx_out_confirmation_number.copy_from_slice(confirmation.as_ref());
-        Ok(mc_util_serial::encode(&tx_out))
-    })
+    unimplemented!("TransactionBuilder::add_output_with_fog_hint_address was removed, please use add_change_output");
```

Comment:
Could you flag this with `#[deprecated]`---ideally the manually constructed headers are also flagged with `__attribute__(deprecated("Use mc_transaction_builder_add_change_output"))`, but you'd need some cross-platform macros since `deprecated` is apparently a GNU extension.

Reply:
Why is `#[deprecated]` better than putting `unimplemented!()` ?

By putting `unimplemented!()` I have done more than just deprecate it, I have actually broken code that is calling it. This is necessary because we made the transaction builder API it was calling private and not public anymore.

Reply:
Because unimplemented is a runtime file, deprecated is a compiler warning.

Reply:
yes, i'm not attempting to create a deprecation cycle, i'm attempting to remove this function since I don't think Adam will be using it in the next release, and any form of testing will eventually hit this, since every transaction involves a change tx out

Reply:
In review of the first PR, eran suggested to just delete it rather than do a deprecation cycle, if the only consumer is libmobilecoin

Reply:
I guess I'm not sure what making this a runtime error accomplishes, if the goal is to remove the function anyways, I'd much rather break the build and just fix it...

Reply:
@the-real-adammork I'm happy to do whatever is most convenient for you

Reply:
I don't have a preference. Seems like I'll need to make Swift changes either way.

Reply:
iOS/Swift code does not call into this function directly